category/traversable/derive.lean

```diff
@@ -0,0 +1,208 @@
+
+meta def nested_map (f v : expr) : expr → tactic expr
+| t :=
+do t ← instantiate_mvars t,
+   mcond (succeeds $ is_def_eq t v)
+      (pure f)
+      (if ¬ v.occurs (t.app_fn)
+          then do
+            cl ← mk_app ``functor [t.app_fn],
+            _inst ← mk_instance cl,
+            f' ← nested_map t.app_arg,
+            mk_mapp ``functor.map [t.app_fn,_inst,none,none,f']
+          else fail format!"type {t} is not a functor with respect to variable {v}")
+
+meta def map_field (n : name) (cl f v e : expr) : tactic expr :=
+do t ← infer_type e >>= whnf,
+   if t.get_app_fn.const_name = n
+   then fail "recursive types not supported"
+   else if v.occurs t
+   then do f' ← nested_map f v t,
+           pure $ f' e
+   else
+         (is_def_eq t.app_fn cl >> to_expr ``(comp.mk %%e))
+     <|> pure e
+
+meta def nested_traverse (f v : expr) : expr → tactic expr
+| t :=
+do t ← instantiate_mvars t,
+   mcond (succeeds $ is_def_eq t v)
+      (pure f)
+      (if ¬ v.occurs (t.app_fn)
+          then do
+            cl ← mk_app ``traversable [t.app_fn],
+            _inst ← mk_instance cl,
+            f' ← nested_traverse t.app_arg,
+            mk_mapp ``traversable.traverse [t.app_fn,_inst,none,none,none,none,f']
+          else fail format!"type {t} is not traversable with respect to variable {v}")
+
+meta def traverse_field (n : name) (inst cl f v e : expr) : tactic expr :=
+do t ← infer_type e >>= whnf,
+   if t.get_app_fn.const_name = n
+   then fail "recursive types not supported"
+   else if v.occurs t
+   then do f' ← nested_traverse f v t,
+           pure $ f' e
+   else
+         (is_def_eq t.app_fn cl >> to_expr ``(comp.mk %%e))
+     <|> to_expr ``(@pure _ (%%inst).to_has_pure _ (ulift.up %%e))
+
+meta def seq_apply_constructor : list expr → expr → tactic expr
+| (x :: xs) e := to_expr ``(%%e <*> %%x) >>= seq_apply_constructor xs
+| [] e := return e
+
+meta def fill_implicit_arg' : expr → expr → tactic expr
+| f (expr.pi n bi d b) :=
+if b.has_var then
+do v ← mk_meta_var d,
+   fill_implicit_arg' (expr.app f v) (b.instantiate_var v)
+else return f
+| e _ := return e
+
+meta def fill_implicit_arg (n : name) : tactic expr :=
+do c ← mk_const n,
+   t ← infer_type c,
+   fill_implicit_arg' c t
+
+meta def mk_down (e : expr) : tactic expr :=
+to_expr ``(ulift.down %%e) <|> pure e
+
+meta def map_constructor (c n : name) (f v : expr) (args : list expr) : tactic unit :=
+do g ← target,
+   args' ← mmap (map_field n g.app_fn f v) args,
+   constr ← fill_implicit_arg c,
+   let r := constr.mk_app args',
+   () <$ tactic.exact r
+
+meta def mk_map (type : name) (cs : list name) := do
+   [α,β,f,x] ← tactic.intro_lst [`α,`β,`f,`x],
+   reset_instance_cache,
+   -- x ← get_local `x,
+   xs ← tactic.induction x,
+   -- f ← get_local `f,
+   -- α ← get_local `α,
+   -- β ← get_local `β,
+   () <$ mzip_with'
+      (λ (c : name) (x : name × list expr × list (name × expr)),
+      solve1 (map_constructor c type f α x.2.1))
+      cs xs
```

Comment:
A side note on `intro_lst`: if the induction tactic reverts some variables, it may invalidate the local constants. Reading this code I'm not sure what exactly is being proved, so this might not be a problem. (You could put in a comment stating in brief what type of goal you are solving? I won't insist on this but it would make good code documentation for later revision.)

Reply:
You're right, the whole file is rather poorly documented. I'll add more documentation. This function for instance could probably use a header:

```
/-- derive the `map` definition of a `functor` -/
```

Reply:
Can this function be used to define something like `list.map` as a universe polymorphic function before specializing it to one universe in the functor?

Reply:
Except for the part that I have removed (temporarily) the support for recursion, I think it should be able to handle universe polymorphism. Would you like to have it to generate a stand-alone polymorphic map and only reference it in the instance?

Reply:
I'm preparing another pull request that includes support for recursive types and derives `is_lawful_*`. Are you in favour of making each law a separate (possibly more polymorphic) lemma or just making `map` and `traverse` separate functions?

Reply:
Does `traverse` have an extra-polymorphic standalone version as well? If you have them, then you should state the lemmas in both forms as well - a standalone `map_id`, plus an instantiation of it in the `is_lawful_functor` instance.

Reply:
Very good. For now, they remain just as polymorphic as the instance but I will make all the definitions and lemmas more polymorphic in the upcoming PRs.